File: 11. Top K frequent elements.cpp
/*
 Task: Find the frequency of elements and return the top K frequent elements
*/

#include <bits/stdc++.h>
using namespace std;

vector<int> topKFrequent(vector<int>& a, int k)
{
    unordered_map<int, int> cnt;
    for (int x : a) cnt[x]++;
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;  // (freq, val)
    for (auto& p : cnt)
    {
        pq.push({p.second, p.first});
        if ((int)pq.size() > k)
            pq.pop();
    }
    vector<int> res;
    while (!pq.empty())
    {
        res.push_back(pq.top().second);
        pq.pop();
    }
    reverse(res.begin(), res.end());
    return res;
}

int main()
{
    vector<int> a = {1, 1, 1, 2, 2, 3};
    auto r = topKFrequent(a, 2);
    cout << "Top K: ";
    for (int x : r) cout << x << " ";
    cout << '\n';
    return 0;
}

// Time Complexity: O(n log k)
// Space Complexity: O(n)

File: 11. Top K frequent elements.py
"""
Task: Find the frequency of elements and return the top K frequent elements
Time Complexity: O(n log k)
Space Complexity: O(n)
"""

import heapq
from collections import Counter

def top_k_frequent(a,k):
    cnt = Counter(a)
    pq = []
    for val,f in cnt.items():
        heapq.heappush(pq, (f, val))
        if len(pq)>k: heapq.heappop(pq)
    res = [heapq.heappop(pq)[1] for _ in range(len(pq))]
    return res[::-1]

print('Top K:', top_k_frequent([1,1,1,2,2,3],2))
